### PR TITLE
add file.stat()

### PR DESCRIPTION
**Pull request by ripienaar**

### CODE REVIEW

**Comment by ripienaar** [comment]
quick attempt at adding #80, please let me know if you need any changes or a different approach.

File: compiler/stdlib/os_file.go [hunk 1]
```diff
@@ -66,6 +67,30 @@ func makeOSFile(file *os.File) *objects.ImmutableMap {
 					return &objects.Int{Value: res}, nil
 				},
 			},
+			// stat() => imap(fileinfo)
```
[the remark this answers is not in the document]
Comment (ripienaar) [reply compiler/stdlib/os_file.go]: yeah, see my comment on the ticket :)

File: compiler/stdlib/os_file.go [hunk 2]
```diff
@@ -66,6 +67,30 @@ func makeOSFile(file *os.File) *objects.ImmutableMap {
...
+			"stat": &objects.UserFunction{
+				Value: func(_ ...objects.Object) (ret objects.Object, err error) {
```
[the remark this answers is not in the document]
Comment (ripienaar) [reply compiler/stdlib/os_file.go]: ok

File: compiler/stdlib/os_file.go [hunk 3]
```diff
@@ -66,6 +67,30 @@ func makeOSFile(file *os.File) *objects.ImmutableMap {
...
+					stat, err := file.Stat()
+					if err != nil {
+						return wrapError(err), nil
+					}
+
+					fstat := map[string]interface{}{
+						"name":      stat.Name(),
+						"mtime":     stat.ModTime().Unix(),
```
[the remark this answers is not in the document]
Comment (ripienaar) [reply compiler/stdlib/os_file.go]: hmm, i will take another look - it was saying unsupported type, perhaps my code wasnt up to date

File: compiler/stdlib/os_file.go [hunk 4]
```diff
@@ -66,6 +67,30 @@ func makeOSFile(file *os.File) *objects.ImmutableMap {
...
+						"directory": stat.IsDir(),
+						"size":      stat.Size(),
+						"mode":      fmt.Sprintf("%04o", stat.Mode()),
```
[the remark this answers is not in the document]
Comment (ripienaar) [reply compiler/stdlib/os_file.go]: int is surprisingly pointless, in this case the int of 0644 would be 420 - its very hard to work with that in tengo hence the choice to add this, perhaps i should add mode thats int and unix_mode thats this string?

File: compiler/stdlib/os_file.go [hunk 5]
```diff
@@ -66,6 +67,30 @@ func makeOSFile(file *os.File) *objects.ImmutableMap {
...
+					}
+
+					o, err := objects.FromInterface(fstat)
```
[the remark this answers is not in the document]
Comment (ripienaar) [reply compiler/stdlib/os_file.go]: Mind expanding on why? I had this and it seems like a whole lot of code that this utility does for us, i guess its slower though this way

[the remark this answers is not in the document]
Comment (d5) [reply compiler/stdlib/os_file.go]: Yeah it's slower. I would use `FromInterface` if there are some values of "unknown" types, but, in this case, all value types are known already.

File: compiler/stdlib/os_file.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (d5) [reply compiler/stdlib/os_file.go]: Tengo parse octal literal `0644` as Int. So you can do something like `state.mode == 0644` when it's Int.

[the remark this answers is not in the document]
Comment (ripienaar) [reply compiler/stdlib/os_file.go]: ah very nice :) ok, I'll rework this tomorrow and maybe add readall etc

thanks for the feedback

**Comment by ripienaar** [comment]
let me know how this looks, if good I'll open a followup for readall

**Comment by d5** [comment]
Yeah, this is good!! 

1. Can you add a test for `os.stat`? (See [text_test](https://github.com/d5/tengo/blob/master/compiler/stdlib/text_test.go) if you need examples)
2. Please add `os.lstat` too. Should be easy.
3. Doc: please add FileInfo (like [ProcessState](https://github.com/d5/tengo/blob/master/docs/stdlib-os.md#processstate)), then update to `stat(string) => FileInfo`

**Comment by ripienaar** [comment]
Addressed the tests and docs - prefer other functions in a separate PR

File: docs/stdlib-os.md [hunk 6]
```diff
@@ -131,6 +133,14 @@ cmd := exec.command("echo", ["foo", "bar"])
 output := cmd.output()
 ```
 
+## FileInfo
+
+- `name`: name of the file the info describes
+- `mtime`: time the file was last modified
+- `size`: file size in bytes
+- `mode`: file permissions as in int, comparable to octal permissions
+- `directory`: boolean indicating if the file is a directory
```
Comment (ripienaar) [inline docs/stdlib-os.md]: wasnt sure how you want these as the other examples only show methods so just tried this to see

**Pull request merged**